tka/aum.go

```diff
@@ -0,0 +1,256 @@
+		}
+	}
+
+	// TODO(tom): Validate State once a future PR brings in that type.
+
+	switch a.MessageKind {
+	case AUMAddKey:
+		if a.Key == nil {
+			return errors.New("AddKey AUMs must contain a key")
+		}
+		if a.KeyID != nil || a.DisablementSecret != nil || a.State != nil || a.Votes != nil || a.Meta != nil {
+			return errors.New("AddKey AUMs may only specify a Key")
+		}
+	case AUMRemoveKey:
+		if len(a.KeyID) == 0 {
+			return errors.New("RemoveKey AUMs must specify a key ID")
+		}
+		if a.Key != nil || a.DisablementSecret != nil || a.State != nil || a.Votes != nil || a.Meta != nil {
+			return errors.New("RemoveKey AUMs may only specify a KeyID")
+		}
+	case AUMUpdateKey:
+		if len(a.KeyID) == 0 {
+			return errors.New("UpdateKey AUMs must specify a key ID")
+		}
+		if a.Meta == nil && a.Votes == nil {
+			return errors.New("UpdateKey AUMs must contain an update to votes or key metadata")
+		}
+		if a.Key != nil || a.DisablementSecret != nil || a.State != nil {
+			return errors.New("UpdateKey AUMs may only specify KeyID, Votes, and Meta")
+		}
+	case AUMCheckpoint:
+		if a.State == nil {
+			return errors.New("Checkpoint AUMs must specify the state")
+		}
+		if a.KeyID != nil || a.DisablementSecret != nil || a.Key != nil || a.Votes != nil || a.Meta != nil {
+			return errors.New("Checkpoint AUMs may only specify State")
+		}
+	case AUMDisableNL:
+		if len(a.DisablementSecret) == 0 {
+			return errors.New("DisableNL AUMs must specify a disablement secret")
+		}
+		if a.KeyID != nil || a.State != nil || a.Key != nil || a.Votes != nil || a.Meta != nil {
+			return errors.New("DisableNL AUMs may only a disablement secret")
+		}
+	}
+
+	return nil
+}
+
+// Serialize returns the given AUM in a serialized format.
+func (a *AUM) Serialize() []byte {
+	// Why CBOR and not something like JSON?
+	//
+	// The main function of an AUM is to carry signed data. Signatures are
+	// over digests, so the serialized representation must be deterministic.
+	// Further, experience with other attempts (JWS/JWT,SAML,X509 etc) has
+	// taught us that even subtle behaviors such as how you handle invalid
+	// or unrecognized fields + any invariants in subsequent re-serialization
+	// can easily lead to security-relevant logic bugs. Its certainly possible
+	// to invent a workable scheme by massaging a JSON parsing library, though
+	// profoundly unwise.
+	//
+	// CBOR is one of the few encoding schemes that are appropriate for use
+	// with signatures and has security-conscious parsing + serialization
+	// rules baked into the spec. We use the CTAP2 mode, which is well
+	// understood + widely-implemented, and already proven for use in signing
+	// assertions through its use by FIDO2 devices.
+	out := bytes.NewBuffer(make([]byte, 0, 128))
+	encoder, err := cbor.CTAP2EncOptions().EncMode()
+	if err != nil {
+		// Deterministic validation of encoding options, should
+		// never fail.
+		panic(err)
+	}
+	if err := encoder.NewEncoder(out).Encode(a); err != nil {
```

Comment:
This should be using [`cbor.CanonicalEncOptions`](https://pkg.go.dev/github.com/fxamacker/cbor/v2#CanonicalEncOptions), right?

Reply:
`CanonicalEncOptions` is less strict than the CTAP2 variety, which is also battle-tested for use in signatures by FIDO2 keys. Also the 'Canonical CBOR' rules allow tags, which would provide more valid encodings for a given data than we want.

Reply:
Ah, got it. I couldn't tell from the name of "CTAP2" that this is a form of canonicalization.